modules/programs/ncmpcpp.nix

```diff
@@ -0,0 +1,142 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) types mkIf mkEnableOption mkOption;
+
+  cfg = config.programs.ncmpcpp;
+
+  mpdCfg = config.services.mpd;
+
+  mpdMusicDir = if cfg.mpdMusicDir == null then
+    mpdCfg.musicDirectory
+  else
+    toString cfg.mpdMusicDir;
+
+  renderSettings = settings:
+    lib.concatStringsSep "\n" (lib.mapAttrsToList renderSetting settings);
+
+  renderSetting = name: value: "${name}=${renderValue value}";
+
+  renderValue = option:
+    rec {
+      int = toString option;
+      bool = if option then "yes" else "no";
+      string = option;
+    }.${builtins.typeOf option};
+
+  renderBindings = bindings:
+    lib.concatStringsSep "\n" (map renderBinding bindings);
+
+  renderBinding = { key, command }:
+    lib.concatStringsSep "\n  "
+    ([ ''def_key "${key}"'' ] ++ maybeWrapList command);
+
+  maybeWrapList = xs: if lib.isList xs then xs else [ xs ];
+
+  valueType = types.oneOf [ types.bool types.int types.str ];
+
+  bindingType = types.submodule ({ name, config, ... }: {
+    options = {
+      key = mkOption {
+        type = types.str;
+        description = "Key to bind.";
+        example = "j";
+      };
+
+      command = mkOption {
+        type = types.either types.str (types.listOf types.str);
+        description = "Command or sequence of commands to be executed.";
+        example = "scroll_down";
+      };
+    };
+  });
+
+in {
+  meta.maintainers = with lib.maintainers; [ olmokramer ];
+
+  options.programs.ncmpcpp = {
+    enable =
+      mkEnableOption "ncmpcpp - An ncurses Music Player Daemon (MPD) client";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.ncmpcpp;
+      defaultText = "pkgs.ncmpcpp";
+      description = "Package providing the <code>ncmpcpp</code> command.";
+      example = "pkgs.ncmpcpp.override { ... }";
+    };
+
+    mpdMusicDir = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = ''
+        Value of the <code>mpd_music_dir</code> option. The value of
+        services.mpd.musicDirectory is used if set to <literal>null</literal>.
+      '';
+      example = "~/music";
+    };
+
+    settings = mkOption {
+      type = types.attrsOf valueType;
+      default = { };
+      description = ''
+        Attrset from name of a setting to its value. For available options
+        See
+        <citerefentry>
+          <refentrytitle>ncmpcpp</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry>
+      '';
+      example = lib.literalExample ''
+        {
+          ncmpcpp_directory = "~/.local/share/ncmpcpp";
+        }
+      '';
+    };
+
+    bindings = mkOption {
+      type = types.listOf bindingType;
```

Comment:
Perhaps could be `with types; attrsOf (either str (listOf str))`? I.e., the example would be
```nix
{
  "j" = "scroll_down";
  "k" = "scroll_up";
  "J" = [ "select_item" "scroll_down" ];
  "K" = [ "select_item" "scroll_up" ];
}
```

Reply:
That's what I tried initially, but you can bind the same key multiple times in `~/.config/ncmcpcpp/bindings`. For example, I have this in my config:

```
def_key "x"
  delete_playlist_items
def_key "x"
  delete_browser_items
def_key "x"
  delete_stored_playlist
```

Reply:
Fair enough.